pyoozie/workflow.py

```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2017 "Shopify inc." All rights reserved.
+# Use of this source code is governed by a MIT-style license that can be found in the LICENSE file.
+import typing  # pylint: disable=unused-import
+
+from pyoozie import tags  # pylint: disable=unused-import
+
+
+class AbstractAction(object):
+    pass
+
+
+class Action(AbstractAction):
+    """Concrete action to execute (takes an action tag, e.g. pyoozie.Shell, pyoozie.Email. etc.)"""
+
+    def __init__(self, action_tag, on_error=None):
+        # type: (tags.Shell, AbstractAction) -> None
+        self.__action_tag = action_tag
+        self.__on_error = on_error
+
+
+class Sequence(AbstractAction):
+    """Set of actions to execute sequentially (implemented by chaining actions and 'OK' transitions)"""
+
+    def __init__(self, *actions, **kwargs):
+        # type: (typing.Tuple[AbstractAction, ...], typing.Dict[str, AbstractAction]) -> None
+        self.__on_error = kwargs.get('on_error')
+        self.__actions = actions
```

Comment:
It's probably best to make a copy of the `actions` (and note it in the docstring) so that people can't change them later. Making a copy is also closer to the semantics of what's actually going to happen in the workflow.